### ISSUE TITLE
Outbound traffic to Dex port is banned in corp network

### ISSUE THREAD

**Issue by Jeffwan** [kubeflow/kubeflow#3826]
/kind feature

**Why you need this feature:**
[Is your feature request related to a problem? Please describe in details]
LB listens on 5556 and forward traffic to Dex pod. The problem is some of the corp network ban outbound traffic to non 80 443 ports. It means this solution is not working well in some of the corp network. When we access hostname or ip of LoadBalancer, redirect to dex hangs there and then timeout eventually. 

**Describe the solution you'd like:**
[A clear and concise description of what you want to happen.]

Propose to use `/dex` instead of `:5556/dex` and use one istio-gateway. But after I make the change and have the test, I notice the EnvoyFilter 
https://github.com/kubeflow/kubeflow/blob/d74d12c487c599e0d90cf498bdbbac901330e6e2/deployment/existing/auth_oidc/envoy-filter.yaml#L28-L30

checks all traffic to 443, what we want is to exclude all the traffic to `/dex`. Otherwise, we get recursively forwards between auth service and gateway. 
1. Authentication service will read `/dex/.well-known/openid-configuration`
2. End user will also need to access `/dex/login`

Anyone knows how to exclude path from authFilter. I didn't find a way to do that now. 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Comment by Jeffwan** [comment]
@yanniszark

**Comment by yanniszark** [comment]
Hi @Jeffwan !
This is indeed a limitation for some corp networks and we are working on lifting it.
The solution you describe is a step in the right direction, however it still wouldn't work by itself.
I've opened https://github.com/kubeflow/kubeflow/issues/3839 for tracking the solution.

**Comment by cestum** [comment]
@Jeffwan 
While [#3839](https://github.com/kubeflow/kubeflow/issues/3839) is being fixed, I found a way to host dex in 443/80 instead of 5556 port by editing envoy filter. Dex is hosted separately in below example, one can still use kubeflow endpoint with separate context root for dex (https://kubeflow.example.com/dex)

```
export OIDC_ENDPOINT=https://dex.example.com
export KUBEFLOW_ENDPOINT=https://kubeflow.example.com
```

`kfctl init kubeflow --config=kfctl_existing_arrikto.yaml -V`

- remove 5556 port rule from istio-noauth.yaml LoadBalancer (here deployment/existing/istio/)

- edit deployment/existing/auth_oidc/gateway.yaml 
remove port 5556 and change the dex virtualservice 
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: kubeflow-gateway
  namespace: kubeflow
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "*"
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: auth-virtual-services
  namespace: kubeflow
spec:
  hosts:
  - dex.example.com
  gateways:
  - kubeflow-gateway
  http:
  - route:
     - destination:
          port:
            number: 5556
          host: dex.kubeflow.svc.cluster.local
```

- edit deployment/existing/auth_oidc/envoy-filter.yaml

```
spec:
  workloadLabels:
    istio: ingressgateway
  filters:
  - filterConfig:
      inlineCode: |
        function authorize(request_handle, cluster, host, path)
          local request_url = "http://"..request_handle:headers():get(":authority")..request_handle:headers():get(":path")
          headers, body = request_handle:httpCall(
            cluster,
            {
              [":method"] = "GET",
              [":path"] = path,
              [":authority"] = request_handle:headers():get(":authority"),
              ["X-Auth-Request-Redirect"] = request_url,
            },
            nil,
            10000)
          return headers, body
        end
        function string.starts(String,Start)
           return string.sub(String,1,string.len(Start))==Start
        end

        function envoy_on_request(request_handle)
          local cluster = "outbound|8080||authservice.istio-system.svc.cluster.local"
          local path= request_handle:headers():get(":path")
          local host= request_handle:headers():get(":authority")
          
          --passthrough dex requests
          if host == "dex.example.com" then
            return
          end
          request_handle:logWarn("Handling host: ".. host .. " - " .. path);

          if host == "kubeflow.example.com" then
            -- oauth flow callback with auth code
            if string.starts(path,"/login/oidc?code=") then
              -- use ambassador oauth service to get jwt token from auth code
              headers, body = authorize(request_handle, cluster, host, path)
              request_handle:respond(headers,body)
            end
            -- other conditions...
          end
          --other calls just forward request to ambassador oauth service
          cookie = request_handle:headers():get("Cookie");
          local headers, body = request_handle:httpCall(
            cluster,
            {
              [":method"] = "GET",
              [":path"] = "/",
              [":authority"] = host,
              ["Cookie"] = cookie
            },
            nil,
            10000)
          local status
          for header, value in pairs(headers) do
            if header == ":status" then
               status = value
            end
          end
          --handle redirects
          if status == "302" then
            request_handle:respond(headers,body)
          end
        end
    filterName: envoy.lua
    filterType: HTTP
    insertPosition:
      index: FIRST
    listenerMatch:
      listenerType: GATEWAY
```


kfctl generate and apply...

### Tests
`https://dex.example.com/.well-known/openid-configuration` should shows openid config
`https://kubeflow.example.com/` should redirect to dex endpoint

**Comment by jbottum** [comment]
@Jeffwan could this be changed to a question, and labeled as a documentation issue?

**Comment by yanniszark** [comment]
This is resolved because of the way we setup auth in v0.7.
/close
